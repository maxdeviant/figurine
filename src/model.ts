import produce from 'immer'
import { Lens, Lensed } from './lens'

export type Transform<T> = (model: T) => void

export interface Model<T> {
  /**
   * Runs the given transform and returns a copy of the model after applying the transformation.
   *
   * @param transform The transform to apply to the model.
   */
  with(transform: Transform<T>): T & Model<T>

  /**
   * Views the model through the given lens.
   *
   * @param lens The lens through which to view the model.
   */
  throughLens<TLens>(lens: Lens<T, TLens>): TLens & Lensed<TLens, T & Model<T>>
}

export type ModelInstance<T> = Readonly<T & Model<T>>

export const Model = <T>(props: T): ModelInstance<T> => ({
  ...(props as any),
  with(transform: Transform<T>) {
    return produce(transform)(this as any)
  },
  throughLens<TLens>(lens: Lens<T, TLens>) {
    return lens(this)
  }
})

export const makeModel = <T>(defaultProps?: T) => (
  props: T
): ModelInstance<T> =>
  defaultProps
    ? (Model({ ...(defaultProps as any), ...(props as any) }) as any)
    : Model(props)
