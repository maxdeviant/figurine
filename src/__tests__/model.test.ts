import { Model, makeModel } from '../model'

describe('Model', () => {
  it('creates a model instance', () => {
    const user = Model({
      name: 'Johnny Depp',
      dateOfBirth: new Date('1963-06-09')
    })

    expect(user).toMatchSnapshot()
  })
})

describe('Model.with', () => {
  it('immutably updates the model', () => {
    const user = Model({
      name: 'Brad Pitt'
    })

    const updatedUser = user.with(user => {
      user.name = 'Bradley Cooper'
    })

    expect(user === updatedUser).toBe(false)
    expect(user.name).toBe('Brad Pitt')
    expect(updatedUser.name).toBe('Bradley Cooper')
  })

  it('is chainable', () => {
    type Post = {
      title: string
      views: number
    }

    const post = Model<Post>({
      title: 'Hello, Sailor!',
      views: 0
    })

    const incrementViews = (post: Post) => {
      post.views++
    }

    const updatedPost = post
      .with(incrementViews)
      .with(incrementViews)
      .with(incrementViews)

    expect(post === updatedPost).toBe(false)
    expect(updatedPost.views).toBe(3)
  })
})

describe('makeModel', () => {
  it('returns a function for creating strongly-typed models', () => {
    const Superhero = makeModel<{
      name: string
    }>()

    const batman = Superhero({
      name: 'Batman'
    })

    const superman = Superhero({
      name: 'Superman'
    })

    expect(batman.name).toBe('Batman')
    expect(superman.name).toBe('Superman')
  })

  it('uses the default values for the models', () => {
    const Animal = makeModel<{
      name: string
      age?: number | '???'
    }>({
      name: '',
      age: '???'
    })

    const rover = Animal({
      name: 'Rover'
    })

    expect(rover.name).toBe('Rover')
    expect(rover.age).toBe('???')
  })
})
