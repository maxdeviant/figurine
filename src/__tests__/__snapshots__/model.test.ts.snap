// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model creates a model instance 1`] = `
Object {
  "dateOfBirth": 1963-06-09T00:00:00.000Z,
  "name": "Johnny Depp",
  "throughLens": [Function],
  "with": [Function],
}
`;
